src/handlers/cors.js

```diff
@@ -1,17 +1,79 @@
-module.exports = function corsHandler({ allowedOrigins = ['*'] }) {
+module.exports = function corsHandler(
+  {
+    allowedOrigins = ['*'],
+    allowedMethods = ['GET','PUT','POST','PATCH','DELETE','HEAD','OPTIONS'],
+    allowCredentials = true,
+    allowedHeaders = ['Content-Type'],
+    allowedExposeHeaders = ['WWW-Authenticate', 'Server-Authorization'],
+    maxAge = '600',
+    optionsSuccessStatus = 204,
+    terminatePreflight = false
+  }) {
   return async (ctx, next) => {
-    const { origin } = ctx.request.headers;
-
-    if (origin) {
-      if (allowedOrigins[0] === '*' || allowedOrigins.indexOf(origin) !== -1) {
-        ctx.set('Access-Control-Allow-Credentials', 'true');
-        ctx.set('Access-Control-Allow-Origin', origin);
-        ctx.set('Access-Control-Allow-Methods', 'GET,PUT,POST,PATCH,DELETE,HEAD,OPTIONS');
-        ctx.set('Access-Control-Allow-Headers', 'Content-Type');
-        ctx.set('Access-Control-Expose-Headers', 'WWW-Authenticate,Server-Authorization');
+    const method = ctx.request.method
+    const origin = ctx.request.headers.origin
+    const requestHeaders = ctx.request.headers['access-control-request-headers']
+
+
+    if (method === 'OPTIONS') {
+      configureOrigin(ctx, origin, allowedOrigins);
+      configureCredentials(ctx, allowCredentials);
+      configureMethods(ctx, allowedMethods);
+      configureAllowedHeaders(ctx, requestHeaders, allowedHeaders);
+      configureMaxAge(ctx, maxAge);
+      configureExposedHeaders(ctx, allowedExposeHeaders);
+      if (terminatePreflight) {
```

Comment:
With all the options I tried to set a default that will maintain backwards compatibility so when people upgrade it will not break their code.
Since currently this plugin always pass the request to the next plugin I set this to false.
I think a better default would be to set this to true, but I didn't want to break anyone.

Reply:
Will setting this to true potentially return false options headers? If the origin doesn't support a verb the cors handler would know this, right?

Reply:
I don't think the worker has any way telling what is supported by the origin. 
The developer will have to configure the CORs handler to return the right headers based on their own case. 
This is the situation also today. For example, your current code returns `Access-Control-Allowed-Methods` header with all possible methods, although the origin might possibly support only a few of them.

Reply:
If the request isn't terminated in the handler it will be passed on the origin, so that it can send for instance a 405 back if the method isn't supported. Think it's good to have the option to terminate preflight requests but also think it's good that it defaults to false, so that it's a conscious decision.